### ISSUE TITLE
[Question] Flow of presentation layer

### ISSUE THREAD

**Issue by m-vellinga** [ShadyBoukhary/flutter_clean_architecture#42]
First of thanks for creating this package! It gave us a lot of food for thought about architecture.  

I have some remarks and I am really interested in your point of view on the matter. I seem to have a different interpretation of the View Controller Presenter flow based of your package.  

To quote some parts from the original blog by Uncle Bob about the "presentation" layer:  

https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html#interface-adapters
> [...]
> It is this layer, for example, that will wholly contain the MVC architecture of a GUI. 
> The Presenters,Views, and Controllers all belong in here. The models are likely just 
> data structures that are passed **from the controllers to the use cases**, and then **back 
> from the use cases to the presenters** and views.
> [...]

https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html#crossing-boundaries
> [...]
> It shows the Controllers and Presenters communicating with the Use Cases in the next layer. 
> Note the flow of control. It begins in the controller, moves through the use case, and 
> then winds up executing in the presenter.
> [...]

So I would expect that an action in the view has following flow: `View => Controller => Usecase => Presenter => View`  

With your implementation the flow is like this: `View => Controller => Presenter => Usecase => (Presenter) => Controller => View`   

Could you shed some light on this?

**Comment by ShadyBoukhary** [comment]
@m-vellinga 

Thank you for your interest in the package!

It does indeed seem that my package's flow of control differs slightly from what Uncle Bob's intention was. I was aware of this difference when I created the package and maintain that I have done it for what I believe is a good reason.

Even though that, technically speaking, the original flow of control, `View => Controller => Usecase => Presenter => View`, is possible in Flutter, it is not practical to implement and would add a great level of complexity to the applications without any tangible benefits.  I will try to explain below the rationale behind my decision by presenting two different approaches.

1. _**The Presenter is contained inside the Controller**_
This approach would add complexity without tangible benefits. That is due to how Flutter's UI works _i.e._ widget tree and a build function. Modifying the widget tree from a `Presenter` would require it to implement several interfaces and mixins such as `ChangeNotifier`, as well as have direct access to the `View`'s global keys. Essentially, it would take on all the responsibilities of the `Controller` as it would also need to keep track of the variables used by the `View` to populate the UI. The `Controller` then would be reduced to a simple method call to the `UseCase`, passing it the `Observer` provided in the `Presenter`.  This is why the `Presenter` needs to be contained in the `Controller`. As you can probably already tell, for the `View` to populate widgets using variables, it would then need to call `controller.someVariable`, which would then in turn retrieve some value from the `Presenter`, adding redundancy. The only way the `Presenter` could directly alter the UI is through these variables and the use of a `ChangeNotifier`. This approach would render the `Controller` ultimately **useless**, as it's only job would be to add one empty layer between the View and Presenter merely to satisfy the original flow of control, which is why I decided not to implement it. 

2. **_Both the Controller and the Presenter are contained inside the VIew_**
In this approach, the View populates the UI using variables from the `Presenter`, which would take on the responsibilities of the `Controller` that exist currently in the package. The `View` would call a method from the `Controller` to execute some usecase. However, the usecase requires an `Observer` in order to return the results, which the `Controller` would not be able to provide as it does not have access to the `Presenter`. Of course, a reference to the `Presenter` can be provided to the `Controller` when it is instantiated inside the `View`, but that would defeat the purpose of the architecture, as both the `View` and the `Controller` have access to the `Presenter`. Unnecessary complexity is then added to the architecture and testing becomes more difficult. The original flow of control is achieved but no tangible benefits are reaped. Only complexity is added. The `Controller` remains largely unnecessary.

The original flow of Control works best for Android. I have used the Clean Architecture with Flutter, native iOS, and Android. In my opinion, the original flow of control was created from the Android perspective, as modifying the UI from the Presenter/Control is straightforward and  does not add the level of complexity or result in issues as is the case with Flutter.

I have thought a lot about the relationship between the `Controller` and `Presenter`. Despite the `Presenter` not having much responsibility in my implementation, I have decided to reinterpret its role to one that I think is more fitting. Its role in my implementation is to **_present the data in a format that is usable to the view_**, that would include several things, such as
1. Filtering, sorting, etc
2. Using interface adapters to convert the data into other data structures

Using this approach, I can keep both the `Controller` and `Presenter` in the architecture while maintaining a reasonable and useful role for both of them at the cost of slightly deviating from the original flow of control. I think this is preferable to rendering one of the two obsolete at the benefit of maintaining the original flow of control. 

You might notice that the `Presenter` class in this package is a very insignificant abstract class. That is because I did not want to _require_ the user to implement one. The package here does not dictate that you _need_ the `Presenter` and leaves that up to the user. It is there as a guideline because I think the responsibilities that I have provided above are enough to warrant its existence in most cases. 

In summary, the reason there is a slight difference in the flow of control is that I have decided to reinterpret the role of the `Presenter` to fit the Flutter/Dart API while delegating enough responsibilities for both the `Controller` and `Presenter` to warrant both of their existence.

Let me know if you have any questions.

**Comment by m-vellinga** [comment]
Thanks for the quick response!

A very elaborate and clear answer on the choices you made. I think this information would be a great addition to the readme. Especially the last part starting at `I have thought a lot [...]`

Puts a lot into perspective for me about the right approach in using the package and architecture to it's full potential.
